### PR TITLE
Guess binary model & add conversion script

### PR DESCRIPTION
**Pull request by vhaasteren**
When a binary model is used that PINT does not know about (e.g. T2), an error is thrown. Sometimes the underlying binary model that the user wants to use is easy to guess. We can provide that guess to help the user

This PR provides a utility function that can guess the binary model from a parfile dictionary, and the UnknownBinaryModel error is modified to include a hint for the user as to which binary model they may want to try, such as this one:

`UnknownBinaryModel: Pulsar system/Binary model component T2 is not provided. Perhaps use DDK?`

An error is still thrown, and no attempt is made to convert parameter values. A conversion script is also available, called `t2binary2pint`. This script just uses the `ModelBuilder` class with the option `allow_T2=True`. That will decide on the best binary model, and in the case of DDK it will convert the KOM/KIN parameters

TODO
- [x] When no guess can be made, the error message should not be modified
- [x] Linting
- [x] Unit tests
- [x] Add unit test that checks the two types of messages of UnknownBinaryModel
- [x] Script to make the conversion (just like TCB -> TDB)
- [x] Refactor choose_model
- [x] Test whether converted models are correct
- [x] Test the t2binary2pint script (would solve all uncovered lines)

### CODE REVIEW

**Comment by dlakaplan** [comment]
Very nice.  Would it make sense to further build a function to make this change (sort of like the TCB->TDB function) based on the best-guess?  A scripts too?

**Comment by vhaasteren** [comment]
I think such a script would be convenient to have, and it's in line with not supporting TCB but still providing a script. Same philosophy I guess. Sure, I can add that. As far as I understood from the timers, the only transformation I would need to do are for the DDK parameters KOM/KIN, which have a different definition in T2 compared to DDK. Is that right? That's your expertise, @dlakaplan, is that the only transformation that is necessary?

What I need feedback on wrt the PR implementation is the try/except block I introduce in the `choose_model` function. I did this there, because I felt that it was not the job of `self.all_components.search_binary_components(binary)` to make such a guess. However, I now feel that what I wrote is not elegant enough, and it adds complexity to the logic.

**Comment by dlakaplan** [comment]
> I think such a script would be convenient to have, and it's in line with not supporting TCB but still providing a script. Same philosophy. Sure, I can add that. As far as I understood from the timers, the only transformation I would need to do are for the DDK parameters KOM/KIN, which have a different definition in T2 compared to DDK. That's your expertise, @dlakaplan, is that the only transformation that is necessary?

I think so.  There are routines in `models/binary_ddk.py` to do the conversion.

**Comment by vhaasteren** [comment]
> I think so. There are routines in `models/binary_ddk.py` to do the conversion.

Right, I see. But... I also see that upon reading, the KOM/KIN parameters are converted in the BinaryDDK to the AU convention, and SINI is removed. I am now a bit puzzled by that logic. SINI is only there if we are reading in a Tempo2 parfile, which would have KOM/KIN in the IAU convention to begin with. But BinaryDDK needs them in the DT96 convention. Isn't that a bit of a weird state for the parfile to be in?

My current conversion function will remove SINI and convert KOM/KIN.

**Comment by vhaasteren** [comment]
@dlakaplan, I've just put together an initial t2binary2pint script, and I'm realizing that such a parfile will always be TCB, so it's kind of superseding functionality of tcb2tdb. In fact, I've just implemented it in the ModelBuilder class in the same way.

A single tool that converts tempo2 -> PINT would be more general though, and perhaps not something we'd want to support. Shall I just leave it as a separate?

**Comment by vhaasteren** [comment]
Added iteration review tag, because the converted binary model is likely not correct. For `DDK` I'm using the `_convert_kom` and `_convert_kin` functions, so perhaps I'm missing something important?

**Comment by dlakaplan** [comment]
> Added iteration review tag, because the converted binary model is likely not correct. For `DDK` I'm using the `_convert_kom` and `_convert_kin` functions, so perhaps I'm missing something important?
> 
> **Specific question**
> 
> Use the tool `t2binary2pint` from this PR on, say, the PPTA J1713 parfile: `t2binary2pint J1713-ppta.par J1713-converted.par`. The resulting parfile should give a TDB DDK parfile. I don't think it's correctly converted.

What do you think is wrong?

**Comment by vhaasteren** [comment]
> > Added iteration review tag, because the converted binary model is likely not correct. For `DDK` I'm using the `_convert_kom` and `_convert_kin` functions, so perhaps I'm missing something important?
> > **Specific question**
> > Use the tool `t2binary2pint` from this PR on, say, the PPTA J1713 parfile: `t2binary2pint J1713-ppta.par J1713-converted.par`. The resulting parfile should give a TDB DDK parfile. I don't think it's correctly converted.
> 
> What do you think is wrong?

Oh... I don't know why I thought it was wrong actually. The residuals look the same as they are with libstempo/tempo2. And the original parfile has:

```
BINARY         T2
SINI            KIN
#SINI		0.95093080767197021955      
KOM            92.824444242804586412     1  0.95050573178270669761   
KIN            70.946081736425649543     1  0.40200796477828915076   
```

The new one has:
```
BINARY                                DDK
KIN                    109.05391826357435 1 0.40200796477828915
KOM                   -2.8244442428045886 1 0.9505057317827067
# KINIAU                  70.94608173642565
# KOMIAU                  92.82444424280459
```

I looked at the KOMIAU/KINIAU parameters, and somehow decided they were different. Ok, fine, I'll add some tests for the conversion script and mark this PR as completed. It seems fully functional now.

**Comment by vhaasteren** [comment]
Should be ready now

**Comment by dlakaplan** [comment]
Can you just paste a few examples in here?

**Comment by vhaasteren** [comment]
PPTA DR3 J1713 (skipping JUMP parameters):

```
PSRJ           J1713+0747
ELONG          256.6686963862985921269732614 1  0.00000000090876673235   
ELAT           30.7003598118515997148630925 1  0.00000000167397472599   
F0             218.81184039471733346     1  7.691921346252546715e-14 
F1             -4.0839252783054717257e-16 1  5.4594314711576426696e-22
PEPOCH         55636                       
POSEPOCH       55636                       
DMEPOCH        58983                       
DM             15.988527064339802541     1  0.00025307207531452086   
DM1            -6.5890534863252850064e-05 1  0.00000605342119520716   
DM2            -5.4543887888466437085e-06 1  0.00000101281051729864   
PMELONG        5.2676082172334073583     1  0.00055800795406477962   
PMELAT         -3.4414946315630252599    1  0.00115897614003722393   
PX             0.89811696943912794388    1  0.01256210980106707044   
SINI            KIN
#SINI		0.95093080767197021955      
BINARY         T2
PB             67.825136364466631286     1  0.00000219567958252830   
T0             54303.634515361629319     1  0.00018447137504573347   
A1             32.342422610088938466     1  0.00000008013425585979   
OM             176.19952646143242884     1  0.00097913197491892516   
ECC            7.4940884937095751794e-05 1  0.00000000034213864828   
PBDOT          1.9589808646702856654e-14 1  9.4538916445564820213e-14
OMDOT          0.00015396242225124112747 1  0.00006275870760582978   
M2             0.30165635446915292143    1  0.00768779658427406771   
START          53040.943687135087238       
FINISH         59303.777495481269703       
TZRMJD         56172.283864994755          
TZRFRQ         709.57799999999997453    
TZRSITE        pks   
FD1            -1.4425242329551252494e-05 1  0.00000207839278552605   
FD2            3.1872195202268463094e-06 1  0.00000171340474719996   
FD3            1.3563215055025152626e-06 1  0.00000054795329246384   
KOM            92.824444242804586412     1  0.95050573178270669761   
KIN            70.946081736425649543     1  0.40200796477828915076   
TRES           0.248        
EPHVER         5                           
CLK            TT(BIPM2020)
MODE 1
TIMEEPH        IF99
DILATEFREQ     Y
PLANET_SHAPIRO -1
T2CMETHOD      IAU2000B
NE_SW          0.000
CORRECT_TROPOSPHERE  Y
EPHEM          DE436
NTOA           5141
CHI2R          2.1010 5094
```

becomes:

```
# Created: 2024-04-15T18:07:40.081257
# PINT_version: 0.9.8+282.ge23026b2
# User: vhaasteren
# Host: Rutgers-M1-MBPro.local
# OS: macOS-14.4.1-x86_64-i386-64bit
# Python: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:39:40) [Clang 15.0.7 ]
# Format: pint
PSRJ                           J1713+0747
EPHEM                               DE436
CLK                          TT(BIPM2020)
UNITS                                 TDB
START              53040.9436871350872380
FINISH             59303.7774954812697030
TIMEEPH                              FB90
T2CMETHOD                        IAU2000B
DILATEFREQ                              N
DMDATA                                  N
NTOA                                 5141
CHI2R                               2.101 0 5094.0
TRES                                0.248
ELONG                 256.668696386298620 1 0.00000000090876673235
ELAT                   30.700359811851598 1 0.00000000167397472599
PMELONG                 5.267608217233407 1 0.0005580079540647797
PMELAT                 -3.441494631563025 1 0.001158976140037224
PX                      0.898116969439128 1 0.01256210980106707
ECL                              IERS2010
POSEPOCH           55636.0000000000000000
F0                  218.81184039471733346 1 7.691921346252546715e-14
F1              -4.0839252783054717257e-16 1 5.4594314711576426696e-22
PEPOCH             55636.0000000000000000
CORRECT_TROPOSPHERE                         Y
PLANET_SHAPIRO                          Y
NE_SW                                 0.0
SWM                                   0.0
DM                  15.988527064339802541 1 0.00025307207531452086
DM1             -6.5890534863252850064e-05 1 6.05342119520716e-06
DM2             -5.4543887888466437085e-06 1 1.01281051729864e-06
DMEPOCH            58983.0000000000000000
BINARY                                DDK
PB                  67.825136364466631286 1 2.1956795825283e-06
PBDOT              1.9589808646702858e-14 1 9.453891644556482e-14
A1                      32.34242261008894 1 8.013425585979e-08
A1DOT                                 0.0
ECC                 7.494088493709575e-05 1 3.4213864828e-10
EDOT                                  0.0
T0                 54303.6345153616293190 1 0.00018447137504573347
OM                  176.19952646143242884 1 0.00097913197491892516
OMDOT           0.00015396242225124112747 1 6.275870760582978e-05
M2                     0.3016563544691529 1 0.007687796584274068
A0                                    0.0
B0                                    0.0
GAMMA                                 0.0
DR                                    0.0
DTH                                   0.0
KIN                    109.05391826357435 1 0.40200796477828915
KOM                   -2.8244442428045886 1 0.9505057317827067
# KINIAU                  70.94608173642565
# KOMIAU                  92.82444424280459
# SINI                   0.9452117806686158
FD1               -1.4425242329551252e-05 1 2.07839278552605e-06
FD2                3.1872195202268463e-06 1 1.71340474719996e-06
FD3                1.3563215055025152e-06 1 5.4795329246384e-07
TZRMJD             56172.2838649947550000
TZRSITE                               pks
TZRFRQ                            709.578
```

And PPTA DR3 J0613:

```
PSRJ           J0613-0200
ELONG          93.7990080578252241119696642 1  0.00000000527342482862   
ELAT           -25.4071385368251683716661615 1  0.00000001081494991767   
F0             326.60056196727986788     1  5.5168645420374461129e-13
F1             -1.0230273661253295295e-15 1  3.9055869190811176866e-21
PEPOCH         55636                       
POSEPOCH       55636                       
DMEPOCH        58991                       
DM             38.773619269948962945     1  0.00075685011491238711   
DM1            -9.7690099037869001671e-05 1  0.00001151008531354333   
DM2            -1.5085519903698426756e-05 1  0.00000155975486599981   
PMELONG        2.117609094803266377      1  0.00232499275314834832   
PMELAT         -10.305261236969457742    1  0.00541024949649476696   
PX             0.85425650360254235313    1  0.06272691625352466716   
BINARY         T2
PB             1.1985125750770358367     1  0.00000000002467310927   
A1             1.0914441628309673835     1  0.00000002980102471189   
PBDOT          3.679353219966319937e-14  1  3.923498139518176671e-15 
TASC           50315.269491312552546     1  0.00000005893419374513   
EPS1           3.9497878993773303543e-06 1  0.00000005213069600587   
EPS2           3.6092533823433831367e-06 1  0.00000005525232749017   
START          53044.545290457821217       
FINISH         59645.34970053811595        
TZRMJD         56352.38403350918524        
TZRFRQ         1281.7480000000000473    
TZRSITE        pks   
FD1            -2.9824988539258083678e-05 1  0.00000649081817900048   
FD2            2.3749491702132622238e-05 1  0.00000623740134743948   
FD3            -6.3327550387763506748e-06 1  0.00000240603869934920   
TRES           1.406        
EPHVER         5                           
H3             1.5395869138306373323e-07 1  0.00000002204402957066   
H4             2.5457245374629321327e-07 1  0.00000002937911765009   
CLK            TT(BIPM2020)
MODE 1
TIMEEPH        IF99
DILATEFREQ     Y
PLANET_SHAPIRO -1
T2CMETHOD      IAU2000B
NE_SW          0.000
CORRECT_TROPOSPHERE  Y
EPHEM          DE436
NTOA           4927
CHI2R          2.0104 4882
```

becomes

```
# Created: 2024-04-15T18:08:15.692842
# PINT_version: 0.9.8+282.ge23026b2
# User: vhaasteren
# Host: Rutgers-M1-MBPro.local
# OS: macOS-14.4.1-x86_64-i386-64bit
# Python: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:39:40) [Clang 15.0.7 ]
# Format: pint
PSRJ                           J0613-0200
EPHEM                               DE436
CLK                          TT(BIPM2020)
UNITS                                 TDB
START              53044.5452904578212170
FINISH             59645.3497005381159500
TIMEEPH                              FB90
T2CMETHOD                        IAU2000B
DILATEFREQ                              N
DMDATA                                  N
NTOA                                 4927
CHI2R                              2.0104 0 4882.0
TRES                                1.406
ELONG                  93.799008057825219 1 0.00000000527342482862
ELAT                  -25.407138536825169 1 0.00000001081494991767
PMELONG                2.1176090948032664 1 0.0023249927531483485
PMELAT                -10.305261236969457 1 0.005410249496494767
PX                     0.8542565036025423 1 0.06272691625352467
ECL                              IERS2010
POSEPOCH           55636.0000000000000000
F0                  326.60056196727986788 1 5.516864542037446113e-13
F1              -1.0230273661253295295e-15 1 3.9055869190811176866e-21
PEPOCH             55636.0000000000000000
CORRECT_TROPOSPHERE                         Y
PLANET_SHAPIRO                          Y
NE_SW                                 0.0
SWM                                   0.0
DM                  38.773619269948962945 1 0.00075685011491238711
DM1             -9.769009903786900167e-05 1 1.151008531354333e-05
DM2             -1.5085519903698426756e-05 1 1.55975486599981e-06
DMEPOCH            58991.0000000000000000
BINARY                              ELL1H
PB                  1.1985125750770358367 1 2.467310927e-11
PBDOT                3.67935321996632e-14 1 3.923498139518177e-15
A1                     1.0914441628309675 1 2.980102471189e-08
TASC               50315.2694913125525460 1 5.893419374513e-08
EPS1            3.9497878993773303543e-06 1 5.213069600587e-08
EPS2            3.6092533823433831367e-06 1 5.525232749017e-08
# ECC             5.3504704866044009753e-06
# OM                  47.579428609186540298
H3              1.5395869138306373323e-07 1 2.204402957066e-08
H4              2.5457245374629321327e-07 1 2.937911765009e-08
NHARMS                                  7
FD1               -2.9824988539258085e-05 1 6.49081817900048e-06
FD2                2.3749491702132622e-05 1 6.23740134743948e-06
FD3                -6.332755038776351e-06 1 2.4060386993492e-06
TZRMJD             56352.3840335091852431
TZRSITE                               pks
TZRFRQ                           1281.748
```

File: src/pint/models/model_builder.py [hunk 1]
```diff
@@ -134,6 +137,16 @@ def __call__(
             converted to TDB upon read. If "raw", an unconverted malformed TCB
             TimingModel object will be returned.
 
+        allow_T2 : True, False, or str, optional
+            Whether to convert a T2 binary model to an appropriate underlying
+            binary model. Default is False, and will throw an error upon
+            encountering the T2 binary model. If True, the binary model will be
+            converted to the most appropriate PINT-compatible binary model.
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:144]: What does an str value for this parameter do?

File: src/pint/models/model_builder.py [hunk 2]
```diff
@@ -134,6 +137,16 @@ def __call__(
...
+
+        force_binary_model : str, optional
+            When set to some binary model, this will override the binary model
+            set in the parfile. Defaults to None
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:148]: Please give an example. Like, should it be "ELL1" or "BinaryELL1"?

File: src/pint/models/model_builder.py [hunk 3]
```diff
@@ -151,6 +164,8 @@ def __call__(
         convert_tcb = allow_tcb == True
         allow_tcb_ = allow_tcb in [True, "raw"]
 
+        assert allow_T2 in [True, False]
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:167]: The docstring says this can also be an str

Comment (abhisrkckl) [reply src/pint/models/model_builder.py:167]: Isn't it better to say `isinstance(allow_T2, bool)`?

File: src/pint/models/model_builder.py [hunk 4]
```diff
@@ -413,6 +430,16 @@ def choose_model(self, param_inpar):
             Dictionary of the unique parameters in .par file with the key is the
             parfile line. :func:`parse_parfile` returns this dictionary.
 
+        allow_T2 : True, False, or str optional
+            Whether to convert a T2 binary model to an appropriate underlying
+            binary model. Default is False, and will throw an error upon
+            encountering the T2 binary model. If True, the binary model will be
+            converted to the most appropriate PINT-compatible binary model.
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:437]: What does an str value for this parameter do?

File: src/pint/models/model_builder.py [hunk 5]
```diff
@@ -444,11 +471,17 @@ def choose_model(self, param_inpar):
         # the overall control parameters. This will get us the binary model name
         # build the base fo the timing model
         # pint_param_dict, unknown_param = self._pintify_parfile(param_inpar)
-        binary = param_inpar.get("BINARY", None)
-        if binary is not None:
+        if not force_binary_model:
+            binary = param_inpar.get("BINARY", None)
+        else:
+            binary = [force_binary_model]
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:477]: Maybe change this to an if statement?

File: src/pint/models/model_builder.py [hunk 6]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+
+        force_binary_model : str
+            When set to some binary model, this will override the binary model
+            set in the parfile.
+
+        allow_T2 : True, False, or str optional
+            Whether to convert a T2 binary model to an appropriate underlying
+            binary model. Default is False, and will throw an error upon
+            encountering the T2 binary model. If True, the binary model will be
+            converted to the most appropriate PINT-compatible binary model.
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:585]: What does an str value for this parameter do?

File: src/pint/models/model_builder.py [hunk 7]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+        Returns
+        -------
+        str
+            Name of the binary component
+
+        Note
+        ----
+        If the binary model does not have a PINT model (e.g. the T2 model), an
+        error is thrown with the suggested model that could replace it. If an
+        appropriate model cannot be found, no suggestion is given
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:596]: Some time ago, I made a PR (#1701) to give tailored error messages for un-implemented binary models like MSS. How does this PR interact with that?

File: src/pint/models/model_builder.py [hunk 8]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+        """
+        binary = param_inpar["BINARY"][0]
+
+        # Guess what the binary model should be, regardless of binary parameter
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:600]: BINARY parameter?

File: src/pint/models/model_builder.py (hunk 7 above)
Comment (abhisrkckl) [reply src/pint/models/model_builder.py:596]: Please also mention the allow_T2 option.

File: src/pint/models/model_builder.py [hunk 9]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+        binary_model_guesses = guess_binary_model(param_inpar)
+
+        # Allow for T2 model, gracefully
+        if force_binary_model is not None:
+            # TODO: also need to convert here? Not very clean flow here
+            binary = force_binary_model
+        elif binary == "T2" and allow_T2:
+            binary = binary_model_guesses[0]
+            log.info(f"Found T2 binary model. Gracefully converting T2 to: {binary}.")
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:609]: This should be a warning, IMHO. I think this is something the user will want to know.

File: src/pint/models/model_builder.py [hunk 10]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+
+            # Make sure that DDK parameters are properly converted
+            convert_binary_model(param_inpar, force_binary_model=binary)
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:612]: Can this operation throw an exception? If yes, will catching it here and printing a more informative error message be useful?

File: src/pint/models/model_builder.py [hunk 11]
```diff
@@ -532,6 +565,71 @@ def choose_model(self, param_inpar):
...
+
+        try:
+            binary_cp = self.all_components.search_binary_components(binary)
+
+        except UnknownBinaryModel as e:
+            log.error(f"Could not find binary model {binary}")
+
+            log.info(
+                f"Compatible models with these parameters: {', '.join(binary_model_guesses)}."
+            )
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:622]: This can be part of the error message instead of being INFO?

File: src/pint/models/model_builder.py [hunk 12]
```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
...
+        best-guess
+
+    """
+
+    def add_sini(parameters):
+        """If 'KIN' is a model parameter, Tempo2 doesn't really use SINI"""
+        if "KIN" in parameters:
+            return list(parameters) + ["SINI"]
+        else:
+            return list(parameters)
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:941]: Naively, why is this needed? Isn't KIN providing the same information as SINI?

File: src/pint/models/model_builder.py [hunk 13]
```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
...
+
+    all_components = AllComponents()
+    binary_models = all_components.category_component_map["pulsar_system"]
+
+    # Find all binary parameters
+    # TODO: Check whetehr binary_model[:6]=="Binary" ??
+    binary_parameters_map = {
+        binary_model[6:]: add_sini(
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:949]: What does the index 6 signify? Is it OK to hardcode this value?

File: src/pint/models/model_builder.py [hunk 14]
```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
...
+    binary_parameters_map.update({"Isolated": []})
+    all_binary_parameters = {
+        parname for parnames in binary_parameters_map.values() for parname in parnames
+    }
+
+    # Find all parfile parameters
+    all_parfile_parameters = set(parfile_dict.keys())
+
+    # All binary parameters in the parfile
+    parfile_binary_parameters = all_parfile_parameters & all_binary_parameters
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:963]: Is this an intersection? I haven't seen the & operator used this way before.

File: src/pint/models/model_builder.py [hunk 15]
```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
...
+        "BT",
+        "ELL1",
+        "ELL1H",
+        "ELL1k",
+        "DDK",
+        "DDGR",
+        "DDS",
+        "DD",
+    ]
+
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:984]: Maybe add an assertion here about the `binary_model_priority` list including *all* implemented binary models? This way it will throw an error if somebody implements a new timing model and doesn't update this list.

File: src/pint/models/model_builder.py [hunk 16]
```diff
@@ -817,3 +915,149 @@ def get_model_and_toas(
...
+    # Now select the best-guess binary model
+    priority = [bm for bm in binary_model_priority if bm in allowed_binary_models]
+    omitted = allowed_binary_models - set(priority)
+
+    return priority + list(omitted)
+
+
+def convert_binary_model(
+    parfile_dict, convert_komkin=True, drop_ddk_sini=True, force_binary_model=None
+):
```
Comment (abhisrkckl) [inline src/pint/models/model_builder.py:994]: I think this name is a bit misleading. I thought this was converting one binary model to another. Maybe change it to something like `convert_binary_params_dict`?

File: src/pint/models/timing_model.py [hunk 17]
```diff
@@ -4072,7 +4072,15 @@ class UnknownParameter(TimingModelError):
 class UnknownBinaryModel(TimingModelError):
     """Signal that the par file requested a binary model no in PINT."""
```
Comment (abhisrkckl) [inline src/pint/models/timing_model.py:4073]: no in --> is not in

Comment (vhaasteren) [reply src/pint/models/timing_model.py:4073]: Ha, good catch. I did not write that actually, my git lens blames Jing. But I'm updating it 👍

File: src/pint/models/model_builder.py (hunk 16 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:994]: Yes, you are right. I initially intended that function differently, and I simply did not update the name accordingly. I am changing it

File: src/pint/models/model_builder.py (hunk 15 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:984]: Is an assertion at this place the right place for it? I could also move the `binary_model_priority` somewhere more central (any ideas where?), and then do the assertion in the `test_t2binary2pint` file for instance.

File: src/pint/models/model_builder.py (hunk 13 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:949]: Oh, right...  thanks for pointing this out. I have been meaning to ask you. In the parfile we name the binary models just with DD, BT, etc. But in PINT they are called BinaryBT, BinaryDD, etc. The 6 indicates the number of characters in "Binary". What do you think is the right way to do this? Do we always use the same naming convention here? In principle that is not necessary, so this is probably hacky and not the right way to do this.

File: src/pint/models/model_builder.py (hunk 5 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:477]: Looking at it, I need to rewrite this a bit. I don't know why I even keep track of `binary` here in this way. It's only used for the error message. Previously I needed it to keep track of the model, but I have been a bit sloppy here with the editing.

File: src/pint/models/model_builder.py (hunk 1 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:144]: Changed the docstring

File: src/pint/models/model_builder.py (hunk 3 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:167]: Yes, changing

File: src/pint/models/model_builder.py (hunk 4 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:437]: Changing the docstring

File: src/pint/models/model_builder.py (hunk 2 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:148]: Thinking about it, what would be preferred? I'm using it as "ELL1" rather than "BinaryELL1", but how do we want this in PINT?

Comment (abhisrkckl) [reply src/pint/models/model_builder.py:148]: I personally prefer "ELL1" just to reduce typing.

File: src/pint/models/model_builder.py (hunk 13 above)
Comment (abhisrkckl) [reply src/pint/models/model_builder.py:949]: Maybe replace 6 with `len("Binary")`? This will clarify your intent at least.

Comment (abhisrkckl) [reply src/pint/models/model_builder.py:949]: Actually, the binary classes have a `binary_model_name` attribute. This will have the desired strings like ELL1, DD etc.

File: src/pint/models/model_builder.py (hunk 15 above)
Comment (abhisrkckl) [reply src/pint/models/model_builder.py:984]: Yes... The assertion may be better placed in a test script. 

Maybe `binary_model_priority` can be a private global variable (`_binary_model_priority`) within `model_builder.py`?

Comment (vhaasteren) [reply src/pint/models/model_builder.py]: Done

File: src/pint/models/model_builder.py (hunk 13 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py]: Thanks!

File: src/pint/models/model_builder.py (hunk 5 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py]: Rewritten

File: src/pint/models/model_builder.py (hunk 2 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py]: Done

File: src/pint/models/model_builder.py (hunk 6 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py]: Removed from docstring

File: src/pint/models/model_builder.py (hunk 8 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:610]: Changed

File: src/pint/models/model_builder.py (hunk 9 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:618]: Changed

File: src/pint/models/model_builder.py (hunk 10 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py]: The exception should then already be earlier. I added a try/except block there

File: src/pint/models/model_builder.py (hunk 11 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:631]: Done

File: src/pint/models/model_builder.py (hunk 12 above)
Comment (vhaasteren) [reply src/pint/models/model_builder.py:950]: It's because we need to check parameters against what's in the parfile. PINT doesn't use SINI in DDK, so PINT would find SINI and conclude DDK is not a suitable model. This is just a workaround to make sure PINT would match the DDK model for a T2 model

**Comment by abhisrkckl** [comment]
Also, please add the `allow_T2` and `force_binary_model` options in the `get_model` and `get_model_and_toas` functions. Most people use those instead of `ModelBuilder` directly.

**Comment by vhaasteren** [comment]
Done

**Comment by abhisrkckl** [comment]
Can you please merge nanograv:master? It fixes the macos-latest issue. 

I think this is pretty much ready to go otherwise. I'll merge this once all the tests pass.

**Comment by vhaasteren** [comment]
> Can you please merge nanograv:master? It fixes the macos-latest issue.
> 
> I think this is pretty much ready to go otherwise. I'll merge this once all the tests pass.

Done. I also added one more unit test that checks for an Exception when the binary model cannot be determined
